### ISSUE TITLE
How to add origin to the bottom of label

### ISSUE THREAD

**Issue by dvzhou** [daohoangson/flutter_widget_from_html#548]
```html
<p>对下列加点字的解释，有误的一项是（    ）</p>
<table name="optionsTable">
  <tbody>
    <tr>
      <td width="50%">A． 吾<span class="emphasis">与</span>点也：          赞成</td>
      <td width="50%">B． <span class="emphasis">摄</span>乎大国之间：      威慑</td>
    </tr>
    <tr>
      <td width="50%">C． 春服<span class="emphasis">既</span>成：          已经</td>
      <td width="50%">D． 异乎三子者之<span class="emphasis">撰</span>：    才能</td>
    </tr>
  </tbody>
</table>
```

The above is a piece of HTML content. The text after the emphasis tag in the content needs to add the origin at the bottom of the text. How can this effect be realized?

![image](https://user-images.githubusercontent.com/40446804/124574654-44b62c80-de7d-11eb-98d8-03aa9095d6cc.png)

The picture above shows the effect on PC

Thank you！

**Comment by daohoangson** [comment]
You meant the little dot below some of the words? It doesn't show with just the HTML (see screenshot below). I think you must have some custom CSS rule for those. Can you share the styling too?

<img width="1392" alt="Screen Shot 2021-07-11 at 00 35 55" src="https://user-images.githubusercontent.com/239336/125171747-fdfb6600-e1df-11eb-8eb2-1d1a44047d27.png">

**Comment by dvzhou** [comment]
Yes, you are quite right。In the HTML content above, there is an “emphasis” tag。On PC page，I defined a style about emphasis(see below)。

.emphasis {
  position: relative;
  display: inline-block;
}
.emphasis::after {
  content: '';
  display: block;
  bottom: 0;
  left: 0;
  right: 0;
  height: 3px;
  background: url(./assets/dot.png) repeat-x left center;
}


However, when using it on the mobile terminal, referring to your documentation, I found that I could not achieve similar functions. So I ask for your help. Thank you very much.

**Comment by daohoangson** [comment]
If this is an important visual element, you will need a custom WF. See this sample code:

<details>

<summary>main.dart</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_widget_from_html/flutter_widget_from_html.dart';

const kHtml = '''
<p>对下列加点字的解释，有误的一项是（    ）</p>
<table name="optionsTable">
  <tbody>
    <tr>
      <td width="50%">A． 吾<span class="emphasis">与</span>点也：          赞成</td>
      <td width="50%">B． <span class="emphasis">摄</span>乎大国之间：      威慑</td>
    </tr>
    <tr>
      <td width="50%">C． 春服<span class="emphasis">既</span>成：          已经</td>
      <td width="50%">D． 异乎三子者之<span class="emphasis">撰</span>：    才能</td>
    </tr>
  </tbody>
</table>
''';

class HelloWorldScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(title: Text('Issue 548')),
        body: Padding(
          padding: const EdgeInsets.all(8.0),
          child: HtmlWidget(
            kHtml,
            factoryBuilder: () => _WidgetFactory(),
          ),
        ),
      );
}

class _WidgetFactory extends WidgetFactory {
  @override
  void parse(BuildMetadata meta) {
    if (meta.element.classes.contains('emphasis')) {
      meta.register(BuildOp(
        onTree: (meta, tree) {
          final BuildTree cloned = tree.copyWith(parent: tree.parent);

          tree.replaceWith(WidgetBit.inline(
              tree,
              WidgetPlaceholder(meta).wrapWith((context, child) {
                final tsh = meta.tsb.build(context);
                return Column(
                  children: [
                    buildColumnWidget(
                        meta, tsh, cloned.build().toList(growable: false)),
                    Text('•', style: tsh.styleWithHeight),
                  ],
                );
              })));
        },
      ));
    }
    super.parse(meta);
  }
}
```

</details>

It will look like this:

<img src="https://user-images.githubusercontent.com/239336/126186968-3c1ce97c-30a9-4e62-97d0-1414fcfd7aa1.png" width="300" />

**Comment by dvzhou** [comment]
Yes, it is. It's perfect. Thank you for your patience. This is so useful to me.

But I have another question. I also see many people asking you this question. Why don't you support CSS style? If you support CSS style, it will be more extensible.

**Comment by daohoangson** [comment]
I do plan to add support for CSS stylesheet (the `STYLE` tag) but pseudo selectors (like the `::after` used in your CSS) is very hard to hard to get right.

**Comment by dvzhou** [comment]
Regardless of the way I use it, if I support CSS style, this component will be more powerful than it is now. look forward to.

**Comment by daohoangson** [comment]
That is planned for sure.

**Comment by daohoangson** [comment]
You can follow the progress here https://github.com/daohoangson/flutter_widget_from_html/issues/532

**Comment by dvzhou** [comment]
A new problem, I modified my HTML display logic by referring to the example of adding dots given by you. When I want to add dots to a paragraph, there is a new problem. When there is a return line, the effect of adding dots is not ideal.
[html.txt](https://github.com/daohoangson/flutter_widget_from_html/files/6845488/html.txt)
Attached is my HTML content



class QuestionHtmlFactory extends WidgetFactory {
  @override
  void parse(BuildMetadata meta) {
    // 文字下加点
    if (meta.element.classes.contains('emphasis')) {
      meta.register(
        BuildOp(
          onTree: (meta, tree) {
            final BuildTree cloned = tree.copyWith(parent: tree.parent);
            tree.replaceWith(
              WidgetBit.inline(
                tree,
                WidgetPlaceholder(meta).wrapWith(
                  (context, child) {
                    final tsh = meta.tsb.build(context);
                    List<Widget> dotList = [];
                    for (int i = 0; i < meta.element.text.length; i++) {
                      dotList.add(Text(' • ', style: tsh.styleWithHeight));
                    }
                    return Stack(
                      children: [
                        buildColumnWidget(meta, tsh, cloned.build().toList(growable: false)),
                        Positioned(top: 10.0, child: Row(children: dotList, mainAxisSize: MainAxisSize.min)),
                      ],
                    );
                  },
                ),
              ),
            );
          },
        ),
      );
    }
    super.parse(meta);
  }
}


The above code is modified by referring to the example provided by you. It supports adding a dot at the bottom of each word, but it doesn't work when returning lines.

**Comment by daohoangson** [comment]
Do you have a screenshot of the returning line?

**Comment by dvzhou** [comment]
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-07-20 at 13 09 10](https://user-images.githubusercontent.com/40446804/126265206-812f2799-68a1-4889-821c-adfcfa46f4df.png)
The palindrome text in the picture corresponds to the HTML I provided above. When the palindrome text appears, you can't add points to the palindrome text. I use relative layout here. I know it's caused by this relative layout, but I can't think of a good way to achieve it.

**Comment by daohoangson** [comment]
That happens because `WidgetSpan` is layout as one piece. When the widget is wide like yours, it will be put into the next line. If the emphasis text is long like that, you need to render it as multiple smaller widgets. Take a look at this sample code:

<details>

<summary>main.dart</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_widget_from_html/flutter_widget_from_html.dart';

const kHtml =
    '<p>阅读下面的文言文，完成10-14题。\n冯唐者，其大父赵人。唐为中郎署长，事文帝。文帝辇过，问唐曰：“父老家安在？”唐以实对。文帝曰：“吾居代时，吾尚食监<span style=\"vertical-align:super;font-size:9px;\">①</span>高祛数为我言赵将李齐之贤，战于钜鹿下。今吾每饭，意未尝不在钜鹿也。<span class=\"emphasis\">父</span>知之乎？”唐对曰：“尚不如廉颇、李牧之为将也。”上既闻廉颇、李牧为人，良说，而搏髀曰：“嗟乎!吾独不得廉颇、李牧时为吾将，吾岂忧匈奴哉!”唐曰：“主臣！陛下虽得廉颇、李牧，弗能用也。”上怒，起入<span class=\"emphasis\">禁中</span>。<span style=\"text-decoration:underline;\">良久，召唐让曰：“公奈何众辱我，独无闲处乎</span>？”唐谢曰：“鄙人不知<span class=\"emphasis\">忌讳</span>。”当是之时，匈奴新大入朝郍，杀北地都尉卬。上以胡寇为意，乃卒复问唐曰：“公何以知吾不能用廉颇、李牧也？”唐对曰：“臣大父言，李牧为赵将居边，军市之租皆自用飨士，赏赐决于外，不从中扰也。委任而责成功，故李牧乃得尽其智能，遣选车千三百乘，彀骑万三千，百金之士<span style=\"vertical-align:super;font-size:9px;\">②</span>十万，是以北逐单于，破东胡，灭澹林，西抑强秦，南支韩、魏。当是之时，赵几霸。今臣窃闻魏尚为云中守，其军市租尽以<span class=\"emphasis\">飨</span>士卒，出私养钱，五日一椎牛，飨宾客军吏舍人，是以匈奴远避，不近云中之塞。夫士卒尽家人子，起田中从军，安知尺籍伍符<span style=\"vertical-align:super;font-size:9px;\">③</span>。终日力战，斩首捕虏，上功幕府，一言不相应，文吏以法绳之。其赏不行而吏奉法必用。臣愚，以为陛下法太明，赏太轻，罚太重。且云中守魏尚坐上功首虏差六级，陛下下之吏，削其爵，罚作之。由此言之，陛下虽得廉颇、李牧，弗能用也。臣诚愚，触忌讳，死罪死罪!”<span style=\"text-decoration:underline;\">文帝说</span><span style=\"text-decoration:underline;\">，</span><span style=\"text-decoration:underline;\">是日令冯唐持节赦魏尚，复以为云中守，而拜唐为车骑都尉</span>。七年，景帝立，以唐为楚相，免。<span class=\"emphasis\">武帝立求贤良举冯唐唐时年九十余不能复为官乃以唐子冯遂为郎</span>。\n（节选自《史记·张释之冯唐列传》）\n注：①尚食监：负责皇帝饮食的官员。②百金之士：能有所作为的人。③尺籍伍符：指记载军令、军功的簿籍和军士中各伍互相作保的守则。\n【小题1】 下列对文中画波浪线部分的断句，正确的一项是（   ）\n</p><table name=\"optionsTable\"><tbody><tr><td>A. 武帝立求/贤良举冯唐/唐时年九十余/不能复为官/乃以唐子/冯遂为郎</td></tr><tr><td>B. 武帝立/求贤良/举冯唐/唐时年九十余/不能复为官/乃以唐子冯遂为郎</td></tr><tr><td>C. 武帝立求/贤良举冯唐/唐时年九十余/不能复为官/乃以唐子冯遂为郎</td></tr><tr><td>D. 武帝立/求贤良/举冯唐/唐时年九十余/不能复为官/乃以唐子/冯遂为郎</td></tr></tbody></table>【小题2】 下列文中加点词语的相关内容的解说，不正确的一项是（   ）\n<table name=\"optionsTable\"><tbody><tr><td>A. 父，古代对男子的美称或对老年人的尊称。此处是指汉文帝尊称冯唐为老人家。</td></tr><tr><td>B. 禁中，指皇帝、后妃等居住的地方。因宫中禁卫森严，臣下不得任意出入，故称禁中。</td></tr><tr><td>C. 忌讳，避忌、顾忌。是指因风俗习惯或迷信，禁忌某些认为不吉利的话或举动。</td></tr><tr><td>D. 飨，指用酒食款待人，也泛指对人提供某些东西。但是，所待对象只能是人。</td></tr></tbody></table>【小题3】 下列对原文有关内容的概括和分析不正确的一项是（   ）\n<table name=\"optionsTable\"><tbody><tr><td>A. 冯唐耿直不知忌讳。文帝赞叹李齐行事，冯唐言李齐不如李牧与廉颇；文帝感叹我有廉颇李牧这样的将领就不用担忧匈奴了，冯唐直言文帝有也不能用，丝毫不给文帝面子。</td></tr><tr><td>B. 文帝大度且谦恭。冯唐当面说他有廉颇李牧也不能用，他大怒，拂袖而去，但事后只不过指责冯唐不应当着众人面说而已；匈奴边患重，他又特意询问为何说他不能用这两人。</td></tr><tr><td>C. 冯唐颇有见识。他认为李牧成为良将，除了他诚心对待将士外，还因为作为将军他有奖罚权和战争自主权，所谓“赏赐决于外”和“委任而责成功”，这为后文批评文帝作准备。</td></tr><tr><td>D. 冯唐为魏尚叫屈。冯唐说文帝不能用廉颇李牧，是为魏尚叫屈。魏尚治军如李牧，善待士卒宾客，保一方平安，却因一次战争不利而“下之吏，削其籍，罚作之”,“赏不行而吏奉法必用”。</td></tr></tbody></table>【小题4】 把文中画横线的句子翻译成现代汉语。\n（1）良久，召唐让曰：“公奈何众辱我，独无闲处乎？”\n___________________________________________________________________________________________\n（2）文帝说，是日令冯唐持节赦魏尚，复以为云中守，而拜唐为车骑都尉。\n___________________________________________________________________________________\n【小题5】冯唐就廉颇、李牧之事对汉文帝进行了哪些批评？请简要概括。\n_______________________________________________________________________________________';

class HelloWorldScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(title: Text('Issue 548')),
        body: SingleChildScrollView(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: HtmlWidget(
              kHtml,
              key: UniqueKey(),
              factoryBuilder: () => QuestionHtmlFactory(),
            ),
          ),
        ),
      );
}

class QuestionHtmlFactory extends WidgetFactory {
  @override
  void parse(BuildMetadata meta) {
    if (meta.element.classes.contains('emphasis')) {
      meta.register(BuildOp(
        onTree: (meta, tree) {
          for (final oldBit in tree.bits.toList(growable: false)) {
            // remove existing bits
            oldBit.detach();
          }

          for (final character in meta.element.text.characters) {
            // render a `WidgetSpan` for each character
            tree.add(WidgetBit.inline(tree, Builder(builder: (context) {
              final tsh = meta.tsb.build(context);
              return Column(
                children: [
                  Text(character, style: tsh.styleWithHeight),
                  Text('•', style: tsh.style.copyWith(height: .5)),
                ],
              );
            })));
          }
        },
      ));
    }
    super.parse(meta);
  }
}

``` 

</details>

<img src="https://user-images.githubusercontent.com/239336/126289355-7c3f0f24-0a1d-47b6-b2df-33bfd2465348.png" width="300" />

**Comment by dvzhou** [comment]
Yes, that's it. It's amazing.
